runtime/compiler/optimizer/StringPeepholes.cpp

```diff
@@ -79,6 +79,80 @@
 #include "env/PersistentCHTable.hpp"           // for CLASSHASHTABLE_SIZE, etc
 #include "env/PersistentInfo.hpp"              // for PersistentInfo
 
+/** \brief
+ *     Remove potentialOSRPointHelper calls from start to end inclusive.
+ *
+ *  \param start
+ *     The tree marking the start of the range.
+ *
+ *  \param end
+ *     The tree marking the end of the range.
+ */
+void TR_StringPeepholes::removePotentialOSRPointHelperCalls(TR::TreeTop* start, TR::TreeTop* end)
+   {
+   TR_ASSERT(start->getEnclosingBlock() == end->getEnclosingBlock(), "Does not support range across blocks");
+
+   TR::TreeTop* ttAfterEnd = end->getNextTreeTop();
+
+   if (comp()->getOption(TR_EnableOSR) &&
+       comp()->isOSRTransitionTarget(TR::postExecutionOSR) &&
+       comp()->getOSRMode() == TR::voluntaryOSR)
+      {
+      TR::TreeTop *tt = start;
+      do
+         {
+         TR::Node *osrNode = NULL;
+         if (comp()->isPotentialOSRPoint(tt->getNode(), &osrNode))
+            {
+            if (osrNode->isPotentialOSRPointHelperCall())
+               {
+               dumpOptDetails(comp(), "Remove tt n%dn with potential osr point %p n%dn\n", tt->getNode()->getGlobalIndex(), osrNode, osrNode->getGlobalIndex());
+               TR::TreeTop* prevTT = tt->getPrevTreeTop();
+               TR::TransformUtil::removeTree(comp(), tt);
+               tt = prevTT;
+               }
+            }
+         tt = tt->getNextTreeTop();
+         } while (tt != ttAfterEnd);
+      }
+   }
+
+/** \brief
+ *     Prohibit OSR over range between start and end inclusive.
+ *
+ *     Walk region and ensure all possible OSR transition points has doNotProfile set on their bci
+ *     as they will not be able to reconstruct their stacks due to the optimizations done on them
+ *     or on the surrounding trees.
+ *
+ *  \param start
+ *     The tree marking the start of the range.
+ *
+ *  \param end
+ *     The tree marking the end of the range.
+ */
+void TR_StringPeepholes::prohibitOSROverRange(TR::TreeTop* start, TR::TreeTop* end)
```

Comment:
This also feels like a more general utility. I think it needs a check for voluntary OSR because it is not valid in involuntary OSR.